sys/parsoid.js

```diff
@@ -125,6 +122,51 @@ class ParsoidService {
         };
     }
 
+    _initOpts(opts) {
+        this.options = opts = opts || {};
+        this.parsoidHost = opts.parsoidHost;
+        this.options.stash_ratelimit = opts.stash_ratelimit || 5;
+        this.options.grace_ttl = opts.grace_ttl || 86400;
+        this._blacklist = compileReRenderBlacklist(opts.rerenderBlacklist);
+        if (!opts.parsoidHost) {
+            throw new Error('Parsoid module: the option parsoidHost must be provided!');
+        }
+    }
+
+    _checkStashRate(hyper, req) {
+        if (!hyper.ratelimiter) {
+            return;
+        }
+        if (hyper._rootReq.headers['x-request-class'] !== 'external') {
+            return;
+        }
+        if (!((req.query && req.query.stash) || (req.body && req.body.stash))) {
+            return;
+        }
+        if (hyper._rootReq.headers['x-stash-rate-checked']) {
+            return;
+        }
+        hyper._rootReq.headers['x-stash-rate-checked'] = true;
+        const key = `${hyper.config.service_name}.parsoid_stash.${req.params.domain}.` +
+            `${req.params.title || 'Transform'}.${req.params.revision || 0}|` +
```

Comment:
mmm... do we really wanna include title/revision in a key? This doesn't really limit anything, a malicious client can just iterate over all revisions of all pages

Reply:
Good point, will remove.